client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,260 @@
+//     type Target = dyn IMintModule + Send + Sync + 'static;
+
+//     fn deref(&self) -> &Self::Target {
+//         &*self.0
+//     }
+// }
+
+pub struct MintModule {
+    config: MintModuleConfig,
+}
+
+pub struct MintModuleConfig {
+    some_number: u32,
+}
+
+impl MintModule {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn from_config(config: MintModuleConfig) -> FedimintModule {
+        FedimintModule(Arc::new(Self { config }))
+    }
+}
+
+impl IFedimintModule for MintModule {
+    fn module_key(&self) -> &'static str {
+        "mint"
+    }
+
+    fn generate_output(
+        &self,
+        _amount: Amount,
+        _deterministic_rand: &bitcoin::hashes::sha256::Hash,
+    ) -> (Output, SpendableOutput) {
+        todo!()
+    }
+
+    fn decode_asset(&self, _d: &mut dyn io::Read) -> Result<SpendableOutput, DecodeError> {
+        todo!()
+    }
+}
+
+/// Something that keeps tracks of which nonces (and any other deterministic randomness we need)
+///
+/// Details to be flushed out.
+pub struct DeterministicRandomnessTracker;
+
+#[derive(Default)]
+pub struct Client {
+    modules: BTreeMap<&'static str, FedimintModule>,
+}
+
+#[derive(Debug, Error)]
+pub enum DBError {
+    #[error("not found")]
+    NotFound,
+}
+
+impl Client {
+    pub fn register_module(&mut self, module: FedimintModule) -> &mut Self {
+        self.modules
+            .insert(module.module_key(), module)
+            .expect("Must not register modules with key conflict");
+        self
+    }
+
+    pub fn load_module_assets(&self, _module_key: &'static str) -> Result<Vec<Vec<u8>>, DBError> {
+        todo!();
+    }
+
+    pub fn load_all_assets(&self) -> Result<Vec<SpendableOutput>, DBError> {
+        let mut assets = vec![];
+        for (&module_key, module) in &self.modules {
+            for raw_asset in self.load_module_assets(module_key)? {
+                assets.push(
+                    module
+                        .decode_asset(&mut raw_asset.as_slice())
+                        .expect("corrupted asset in the db"),
+                );
+            }
+        }
+
+        Ok(assets)
+    }
+
+    fn load_deterministic_rng_tracker(&self) -> Result<DeterministicRandomnessTracker, DBError> {
+        todo!()
+    }
+
+    fn start_db_transaction(&self) -> DBTransaction {
+        DBTransaction
+    }
+
+    fn delete_used_spendable_outputs(&self, inputs: &[SpendableOutput]) -> Result<(), DBError> {
+        todo!()
+    }
+
+    fn store_spendable_outputs(&self, new_spendable: &[SpendableOutput]) -> Result<(), DBError> {
+        todo!()
+    }
+}
+
+/// Database transaaction object
+///
+/// Just a placeholder, don't pay too much attention to it.
+struct DBTransaction;
+
+impl DBTransaction {
+    fn commit(&self) -> Result<(), DBError> {
+        todo!()
+    }
+}
+
+#[test]
+fn example_workflow() -> Result<(), Box<dyn std::error::Error>> {
+    let mint_module = MintModule::from_config(MintModuleConfig { some_number: 3 });
```

Comment:
Note that each module can come from a separate package/module. The general purpose code does not have to know about it at all.